### PR TITLE
Added "Zoom All", "Orient Axes", "Reset Axes" buttons

### PR DESCRIPTION
**Pull request by midlik**
# Description
I added "Zoom All", "Orient Axes", "Reset Axes" buttons. These will show when one hovers over the "Reset Camera" button. "Zoom All" does the same as "Reset Camera" (we might drop it).
"Orient Axes"/"Reset Axes" rotate camera to align with PCA/XYZ coordinates. I exposed the same functionality via `PluginCommands` and `plugin.manager.camera` as well.

## Actions

- [x] Added description of changes to the `[Unreleased]` section of `CHANGELOG.md`
- [x] Updated headers of modified files
- [x] Added my name to `package.json`'s `contributors`

### CODE REVIEW

File: src/mol-plugin-state/manager/focus-camera/orient-axes.ts [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+            for (let i = 0; i < unit.elements.length; i++) {
+                const index = unit.elements[i];
+                const compound = unit.model.atomicHierarchy.atoms.label_comp_id.value(index);
+                const element = unit.model.atomicHierarchy.atoms.type_symbol.value(index);
+                if (element !== 'H' && compound !== 'HOH') {
+                    coords.push(x[index], y[index], z[index]);
+                }
+            }
+        }
+    }
```
Comment (dsehnal) [inline src/mol-plugin-state/manager/focus-camera/orient-axes.ts:117]: Usually we use this pattern for accessing the props:

```ts
const loc = StructureElement.Location.create(structure);
const type_symbol = StructureProperties.atom.type_symbol;
for (const unit of units) {
  const { elements } = unit;
  loc.unit = unit;
  for (let i = 0; i < unit.elements.length; i++)
    loc.element = elements[i];
    const ts = type_symbol(loc);
  }
}
```

Can you update your code with that please?

File: src/mol-plugin-state/manager/focus-camera/orient-axes.ts [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    const wrongFrontBack = zCum < 0;
+    const wrongLeftTopRightBottom = wrongFrontBack ? xCum + yCum < 0 : xCum - yCum < 0;
+    if (wrongLeftTopRightBottom && wrongFrontBack) {
+        return ROTATION_MATRICES.rotY180; // flip around Y = around X then Z
+    } else if (wrongFrontBack) {
+        return ROTATION_MATRICES.rotX180; // flip around X
+    } else if (wrongLeftTopRightBottom) {
+        return ROTATION_MATRICES.rotZ180; // flip around Z
+    } else {
+        return Mat3.identity(); // do not flip
```
Comment (dsehnal) [inline src/mol-plugin-state/manager/focus-camera/orient-axes.ts:196]: Cache this matrix too?

File: src/mol-plugin-state/manager/focus-camera/orient-axes.ts [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+            bestScore = score;
+        }
+    }
+    return bestFlip;
+}
+
+/** Measure of how similar two rotation matrices are (dot product of flattened matrices) */
+function mat3Dot(a: Mat3, b: Mat3) {
+    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2] + a[3] * b[3] + a[4] * b[4] + a[5] * b[5] + a[6] * b[6] + a[7] * b[7] + a[8] * b[8];
+}
```
Comment (dsehnal) [inline src/mol-plugin-state/manager/focus-camera/orient-axes.ts:154]: Add this to the `Mat3` namespace instead please.

File: src/mol-plugin-ui/skin/base/components/viewport.scss [hunk 4]
```diff
@@ -25,14 +25,14 @@
     bottom: 0;
     -webkit-user-select: none;
     user-select: none;
-    -webkit-tap-highlight-color: rgba(0,0,0,0);
+    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
     -webkit-touch-callout: none;
     touch-action: manipulation;
 
-    > canvas {
+    >canvas {
```
Comment (dsehnal) [inline src/mol-plugin-ui/skin/base/components/viewport.scss:32]: Please fix the formatting in this CSS file (highlighted by the change view)

File: src/mol-plugin-ui/viewport.tsx [hunk 5]
```diff
@@ -67,9 +72,17 @@ export class ViewportControls extends PluginUIComponent<ViewportControlsProps, V
...
+        this.setState(old => ({ ...old, isCameraResetEnabled: enable }));
+    };
+
     componentDidMount() {
         this.subscribe(this.plugin.events.canvas3d.settingsUpdated, () => this.forceUpdate());
         this.subscribe(this.plugin.layout.events.updated, () => this.forceUpdate());
+        if (this.plugin.canvas3d) {
+            this.subscribe(this.plugin.canvas3d.camera.stateChanged, snapshot => this.enableCameraReset(snapshot.radius !== 0 && snapshot.radiusMax !== 0));
+            // I have no idea what radius and radiusMax mean, but this is the combination that seems to works
+        }
```
Comment (dsehnal) [inline src/mol-plugin-ui/viewport.tsx:85]: Why do you need to disable this?

This is very ineffective given how often the `stateChanged` is called when interacting with the canvas. If you really want to keep this in, at least add a `.pipe` with `throtte/debounce`.

Alternatively, you can just make the reset a non-action if this condition is met.

Comment (midlik) [reply src/mol-plugin-ui/viewport.tsx:85]: I checked this, and found out that `stateChanged` does not fire when using mouse or touchpad to move camera.
There is one exception though: pressing middle button and dragging does trigger `stateChanged`.
Is this the expected behavior?

Comment (dsehnal) [reply src/mol-plugin-ui/viewport.tsx:85]: You are right, we actually use similar code in other places as well, e.g., https://github.com/molstar/molstar/blob/master/src/mol-plugin-ui/left-panel.tsx#L145

File: src/mol-plugin-state/manager/focus-camera/orient-axes.ts (hunk 1 above)
Comment (midlik) [reply src/mol-plugin-state/manager/focus-camera/orient-axes.ts]: Done, and merged the three similar function into one function with options.

File: src/mol-plugin-state/manager/focus-camera/orient-axes.ts (hunk 2 above)
Comment (midlik) [reply src/mol-plugin-state/manager/focus-camera/orient-axes.ts]: Done.

File: src/mol-plugin-state/manager/focus-camera/orient-axes.ts (hunk 3 above)
Comment (midlik) [reply src/mol-plugin-state/manager/focus-camera/orient-axes.ts]: Done, and I found out there is an actual name for this operator :D

File: src/mol-plugin-ui/skin/base/components/viewport.scss (hunk 4 above)
Comment (midlik) [reply src/mol-plugin-ui/skin/base/components/viewport.scss]: Fixed (for some reason my autoformat dislikes the space after >).

File: src/mol-plugin-ui/viewport.tsx (hunk 5 above)
Comment (midlik) [reply src/mol-plugin-ui/viewport.tsx]: I added `.pipe(throttleTime(...))` to all subscriptions to `stateChanged`.

**Review by dsehnal — APPROVED** [review]

**Pull request merged**